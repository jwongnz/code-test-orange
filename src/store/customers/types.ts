export interface Customer {
  id: string;
  firstName: string;
  lastName: string;
  dateOfBirth: Date;
}

export const ADD_CUSTOMER = 'ADD_CUSTOMER';
export const UPDATE_CUSTOMER = 'UPDATE_CUSTOMER';
export const DELETE_CUSTOMER = 'DELETE_CUSTOMER';

export interface AddCustomerAction {
  type: typeof ADD_CUSTOMER;
  payload: Customer
}

export interface UpdateCustomerAction {
  type: typeof UPDATE_CUSTOMER;
  payload: Customer
}


export interface DeleteCustomerAction {
  type: typeof DELETE_CUSTOMER;
  payload: Customer
}

export type ActionTypes = AddCustomerAction | UpdateCustomerAction | DeleteCustomerAction;