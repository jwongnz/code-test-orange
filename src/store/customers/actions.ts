import { ADD_CUSTOMER, DELETE_CUSTOMER, Customer } from './types';
import uuid from 'uuid';

export function addCustomer(customer: Customer) {
  return {
    type: ADD_CUSTOMER,
    payload: {
      ...customer,
      id: uuid.v4()
    }
  };
}

export function deleteCustomer(customer: Customer) {
  return {
    type: DELETE_CUSTOMER,
    payload: customer
  };
}
