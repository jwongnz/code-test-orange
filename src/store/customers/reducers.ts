import { ActionTypes, ADD_CUSTOMER, DELETE_CUSTOMER, Customer } from './types';

const initialState: Customer[] = [
  { id: 'bfc2b7af-6cf3-44bd-a98f-8e2e86790d9a', firstName: 'John', lastName: 'Smith', dateOfBirth: { day: 12, month: 'January', year: 1992 } },
  { id: 'e3493230-9e4a-4899-8ff2-1930f8f60119', firstName: 'Bob', lastName: 'Sushi', dateOfBirth: { day: 20, month: 'March', year: 2001 } }
];

export default function(state = initialState, action: ActionTypes): Customer[] {
  switch(action.type) {
    case ADD_CUSTOMER:
      if (!action.payload)
        return state;

      return [
        ...state,
        action.payload
      ];
    case DELETE_CUSTOMER:
      if (!action.payload)
        return state;

      return state.filter((customer) => {
        return customer.id !== action.payload.id;
      });
    default:
      return state;
  }
}