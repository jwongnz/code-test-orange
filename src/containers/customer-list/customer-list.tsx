import React from 'react';
import { connect } from 'react-redux';
import { AppState } from '../../store';
import { deleteCustomer } from '../../store/customers/actions';
import { Customer } from '../../store/customers/types';
import { Table, TableBody, TableCell, TableHead, TableRow, IconButton } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';
import EditIcon from '@material-ui/icons/Edit';
import { Link } from 'react-router-dom';

interface AppProps {
  customers: Customer[],
  deleteCustomer: typeof deleteCustomer
}

class CustomerList extends React.Component<AppProps> {
  deleteCustomer = (customer: Customer) => {
    this.props.deleteCustomer(customer)
  }

  render() {
    if (!(this.props.customers && this.props.customers.length > 0))
      return null;

    return (
      <Table>
        <TableHead>
          <TableRow>
            <TableCell></TableCell>
            <TableCell>First Name</TableCell>
            <TableCell>Last Name</TableCell>
            <TableCell>Date of Birth</TableCell>
            <TableCell></TableCell>
            <TableCell></TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {this.props.customers.map((customer, index) => (
            <TableRow key={customer.id}>
              <TableCell>{index + 1}</TableCell>
              <TableCell>{customer.firstName}</TableCell>
              <TableCell>{customer.lastName}</TableCell>
              <TableCell>
                {customer.dateOfBirth.toDateString()}
              </TableCell>
              <TableCell>
                <IconButton component={Link} to={`/savecustomer/${customer.id}`} aria-label="Edit">
                  <EditIcon />
                </IconButton>
              </TableCell>
              <TableCell>
                <IconButton aria-label="Delete" onClick={() => this.deleteCustomer(customer)}>
                  <DeleteIcon />
                </IconButton>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    )
  }
}

const mapStateToProps = (state: AppState) => ({
  customers: state.customers
});

const mapDispatchToProps = { deleteCustomer };

export default connect(mapStateToProps, mapDispatchToProps)(CustomerList);
