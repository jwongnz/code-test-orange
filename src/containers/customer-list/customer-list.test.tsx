import React from 'react';
import { CustomerList } from './customer-list';
import { shallow } from 'enzyme';
import { Customer } from '../../store/customers/types';
import toJson from 'enzyme-to-json';
import { TableRow } from '@material-ui/core';

describe('CustomerList', () => {
  let props: any;
  let mockCustomer: Customer;

  beforeEach(() => {
    props = {
      searchTerm: '',
      customers: [
        {id: 'a', firstName: 'John', lastName: 'Smith', dateOfBirth: new Date('1992-05-03')},
        {id: 'b', firstName: 'Dylan', lastName: 'Skywaltzer', dateOfBirth: new Date('1992-05-03')},
        {id: 'c', firstName: 'Bob', lastName: 'Smith', dateOfBirth: new Date('2001-08-14')},
        {id: 'd', firstName: 'Luke', lastName: 'Skywalker', dateOfBirth: new Date('2001-08-14')},
        {id: 'e', firstName: 'Sigmund', lastName: 'Knight', dateOfBirth: new Date('1992-05-03')},
        {id: 'f', firstName: 'Teemo', lastName: 'Mushroom', dateOfBirth: new Date('2001-08-14')},
        {id: 'g', firstName: 'Teeto', lastName: 'Potatoes', dateOfBirth: new Date('2001-08-14')}
      ],
      deleteCustomer: jest.fn()
    }

    mockCustomer = {id: 'abc', firstName: 'John', lastName: 'Smith', dateOfBirth: new Date('1992-05-03')};
  });

  it('renders without crashing', () => {
    shallow(<CustomerList {...props}/>);
  });

  it('renders correctly', () => {
    const wrapper = shallow(<CustomerList {...props}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('search', () => {
    it('should show all customers if no search term', () => {
      const wrapper = shallow(<CustomerList {...props}/>);
      expect(wrapper.find(TableRow)).toHaveLength(8);
    });

    it('shows customers based on search term', () => {
      props.searchTerm = 'Smith';
      const wrapper = shallow(<CustomerList {...props}/>);
      expect(wrapper.find(TableRow)).toHaveLength(3);
    });
  });

  describe('deleteCustomer', () => {
    it('dispatches deleteCustomer action', () => {
      // arrange
      const wrapper = shallow<CustomerList>(<CustomerList {...props}/>);
      const instance = wrapper.instance();

      // act
      instance.deleteCustomer(mockCustomer);
      
      // assert
      expect(props.deleteCustomer).toHaveBeenCalledWith(mockCustomer);
    });
  });
});