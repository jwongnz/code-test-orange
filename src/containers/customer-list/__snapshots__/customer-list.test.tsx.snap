// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerList renders correctly 1`] = `
<WithStyles(ForwardRef(Table))>
  <WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableCell)) />
      <WithStyles(ForwardRef(TableCell))>
        First Name
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Last Name
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Date of Birth
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell)) />
      <WithStyles(ForwardRef(TableCell)) />
    </WithStyles(ForwardRef(TableRow))>
  </WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableBody))>
    <WithStyles(ForwardRef(TableRow))
      key="a"
    >
      <WithStyles(ForwardRef(TableCell))>
        1
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        John
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Smith
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Sun May 03 1992
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Edit"
          component={[Function]}
          to="/savecustomer/a"
        >
          <EditIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Delete"
          onClick={[Function]}
        >
          <DeleteIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableRow))
      key="b"
    >
      <WithStyles(ForwardRef(TableCell))>
        2
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Dylan
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Skywaltzer
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Sun May 03 1992
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Edit"
          component={[Function]}
          to="/savecustomer/b"
        >
          <EditIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Delete"
          onClick={[Function]}
        >
          <DeleteIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableRow))
      key="c"
    >
      <WithStyles(ForwardRef(TableCell))>
        3
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Bob
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Smith
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Tue Aug 14 2001
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Edit"
          component={[Function]}
          to="/savecustomer/c"
        >
          <EditIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Delete"
          onClick={[Function]}
        >
          <DeleteIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableRow))
      key="d"
    >
      <WithStyles(ForwardRef(TableCell))>
        4
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Luke
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Skywalker
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Tue Aug 14 2001
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Edit"
          component={[Function]}
          to="/savecustomer/d"
        >
          <EditIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Delete"
          onClick={[Function]}
        >
          <DeleteIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableRow))
      key="e"
    >
      <WithStyles(ForwardRef(TableCell))>
        5
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Sigmund
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Knight
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Sun May 03 1992
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Edit"
          component={[Function]}
          to="/savecustomer/e"
        >
          <EditIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Delete"
          onClick={[Function]}
        >
          <DeleteIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableRow))
      key="f"
    >
      <WithStyles(ForwardRef(TableCell))>
        6
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Teemo
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Mushroom
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Tue Aug 14 2001
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Edit"
          component={[Function]}
          to="/savecustomer/f"
        >
          <EditIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Delete"
          onClick={[Function]}
        >
          <DeleteIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableRow))
      key="g"
    >
      <WithStyles(ForwardRef(TableCell))>
        7
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Teeto
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Potatoes
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Tue Aug 14 2001
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Edit"
          component={[Function]}
          to="/savecustomer/g"
        >
          <EditIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Delete"
          onClick={[Function]}
        >
          <DeleteIcon />
        </WithStyles(ForwardRef(IconButton))>
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
  </WithStyles(ForwardRef(TableBody))>
</WithStyles(ForwardRef(Table))>
`;
